File: src/problemsolvingpatterns/frequencycounter/anagram/Anagram.js
/*
    Given two strings, write a function to determine if the second string is 
    an anagram of the first. An anagram is a word, phrase, or name formed by rearranging 
    the letters of another, such as cinema, formed from iceman.

validAnagram('', '') // true
validAnagram('aaz', 'zza') // false
validAnagram('anagram', 'nagaram') // true
validAnagram("rat","car") // false) // false
validAnagram('awesome', 'awesom') // false
validAnagram('qwerty', 'qeywrt') // true
validAnagram('texttwisttime', 'timetwisttext') // true


1 - Solution O(n**2)

// if the length of the strings are not equal, return false;
// loop over the first string
    // if current element exists in the second string, remove it from the second string
    // else return false
// return false

2 - Solution O(n)

// if the length of the string are not equal, return false;
// define lookup object
// loop over the first string
    // if the element is the key in the lookup object, increment by one
    // else add it to the object as a key and assign value one.
// loop over the second string
    // if the element is not the key in the lookup object 
    // or if exists, the value is less than or equal to zero
    // then return false
    // else decrement by one
// return true
*/

// 1 - Solution O(n**2)
function validAnagram(str1, str2) {
  if (str1.length !== str2.length) {
    return false;
  }
  for (let el of str1) {
    let correctIndex = str2.indexOf(el);

    if (correctIndex === -1) {
      return false;
    }
    str2 = str2.substring(0, correctIndex) + str2.substring(correctIndex);
  }
  return true;
}

// 2 - Solution O(n)
function validAnagram(str1, str2) {
  if (str1.length !== str2.length) {
    return false;
  }
  const lookup = {};
  for (let el of str1) {
    lookup[el] = ++lookup[el] || 1;
  }
  for (let el of str2) {
    if (!lookup[el]) {
      return false;
    } else {
      lookup[el]--;
    }
  }
  return true;
}

validAnagram('cat', 'act')

File: src/problemsolvingpatterns/frequencycounter/anagram/Anagram.java
package problemsolvingpatterns.frequencycounter.anagram;

import java.util.HashMap;
import java.util.Map;

/**
 * Given two strings, write a function to determine if the second string is 
 * an anagram of the first. An anagram is a word, phrase, or name formed by rearranging 
 * the letters of another, such as cinema, formed from iceman.
 * 
 * Examples:
 * 
 * validAnagram('', '') -> true
 * validAnagram('aaz', 'zza') // false
 * validAnagram('anagram', 'nagaram') // true
 * validAnagram("rat","car") // false) // false
 * validAnagram('awesome', 'awesom') // false
 * validAnagram('qwerty', 'qeywrt') // true
 * validAnagram('texttwisttime', 'timetwisttext') // true
 */
public class Anagram {
  /**
   * Approach #1: Complexity O(n^2)
   * 
   * loop over the each character in the first string
   *  if current char exists i the second string, remove it
   *  else return false
   * return true;
   * 
   * @param str1
   * @param str2
   * @return
   */
  public static boolean validAnagram1(String str1, String str2) {
        for (int i = 0; i < str1.length(); i++) {
      int index = str2.indexOf(str1.charAt(i));
      if (index == -1) {
        return false;
      }
      str2 = str2.substring(0, index) + str2.substring(index + 1);
    }
    return true;
  }

  /**
   * Approach #2: Complexity O(n)
   * 
   * declare map variable to store the frequency of each character in the second string
   * loop over each char in second string
   *  if char exists in the map, increase the value by one
   *  else put it with value 1
   * loop over the second string
   *  if current char doesn't exist in the map or the value is less than or equal to nol, then return false
   *  else decrease the value by one
   * return true
   * 
   * @param str1
   * @param str2
   * @return
   */
  public static boolean validAnagram2(String str1, String str2) {
    if (str1.length() != str2.length()) return false;

    Map<Character, Integer> lookup = new HashMap<>();
    for (int i = 0; i < str2.length(); i++) {
      char ch = str2.charAt(i);
      lookup.put(ch, lookup.getOrDefault(ch, 0) + 1);
    }
    for (int i = 0; i < str1.length(); i++) {
      char ch = str1.charAt(i);
      int count = lookup.getOrDefault(ch, 0);
      if (count <= 0) {
        return false;
      }

      lookup.put(ch, count - 1);
    }
    return true;
  }

  public static void main(String[] args) {
    test("", "", 1);
    test("aaz", "zza", 2);
    test("anagram", "nagaram", 3);
    test("rat", "car", 4);
    test("awesome", "awesom", 5);
    test("qwerty", "qeywrt", 6);
    test("texttwisttime", "timetwisttext", 7);
  }

  public static void test(String str1, String str2, int caseNumber) {
    System.out.println("Case #" + caseNumber);
    System.out.println("'" + str1 + "'" + " and " + "'" + str2 + "'" + ": " + validAnagram2(str1, str2));
  }
}
